File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.java
import java.util.Arrays;

class Solution {
    public int solution(int[][] triangle) {
        int n = triangle.length;
        int[][] dp = new int[n][];
        for (int i = 0; i < n; i++) {
            dp[i] = new int[i+1];
        }
        
        dp[0][0] = triangle[0][0];
        
        for (int i = 1; i < n; i++) {
            for (int j = 0; j < triangle[i].length - 1; j++) {
                int tmp1 = triangle[i][j] + dp[i-1][j];
                dp[i][j] = Math.max(dp[i][j], tmp1);
                
                int tmp2 = triangle[i][j+1] + dp[i-1][j];
                dp[i][j+1] = Math.max(dp[i][j+1], tmp2);

            }
        }
        
        int answer = Arrays.stream(dp[n-1]).max().getAsInt();
        
        return answer;
    }
}

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.py
def solution(triangle):
    n = len(triangle)
    
    dp = [[0] * i for i in range(1, n+1)]
    
    dp[0][0] = triangle[0][0]
    
    for i in range(1, n):
        for j in range(len(triangle[i])-1):
            dp[i][j] = max(dp[i-1][j] + triangle[i][j], dp[i][j])
            dp[i][j+1] = max(dp[i-1][j] + triangle[i][j+1], dp[i][j+1])
    
    answer = max(dp[-1])
    
    return answer